From 2baab97a01a04a1eb3cf9228b85074bb5bad165e Mon Sep 17 00:00:00 2001
From: Ranjani Vaidyanathan <Ranjani.Vaidyanathan@freescale.com>
Date: Mon, 20 Apr 2015 15:14:05 -0500
Subject: [PATCH 0254/1594] MLK-10704 cpufreq:imx6:Ensure that kernel can boot
 with performance governor

commit 89affc80bb21bd9a620410b5661ded3d1eb2cd7c from
git://git.freescale.com/imx/linux-2.6-imx.git

Move the mutex initialization before registering the CPUFREQ driver.
This is required because registering the driver can trigger a call to
change the CPU freq when performance governor is selected as default governor.
Also add rcu_read_lock/unlock around dev_pm_opp_get_opp_count() calls as
specified by the API.

Signed-off-by: Ranjani Vaidyanathan <Ranjani.Vaidyanathan@freescale.com>
---
 drivers/cpufreq/imx6q-cpufreq.c |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/drivers/cpufreq/imx6q-cpufreq.c b/drivers/cpufreq/imx6q-cpufreq.c
index a21cbb7..a63f6b3 100644
--- a/drivers/cpufreq/imx6q-cpufreq.c
+++ b/drivers/cpufreq/imx6q-cpufreq.c
@@ -297,7 +297,9 @@ static int imx6q_cpufreq_probe(struct platform_device *pdev)
 	 * Just, incase the platform did not supply the OPP
 	 * table, it will try to get it.
 	 */
+	rcu_read_lock();
 	num = dev_pm_opp_get_opp_count(cpu_dev);
+	rcu_read_unlock();
 	if (num < 0) {
 		ret = of_init_opp_table(cpu_dev);
 		if (ret < 0) {
@@ -308,7 +310,9 @@ static int imx6q_cpufreq_probe(struct platform_device *pdev)
 		/* Because we have added the OPPs here, we must free them */
 		free_opp = true;
 
+		rcu_read_lock();
 		num = dev_pm_opp_get_opp_count(cpu_dev);
+		rcu_read_unlock();
 		if (num < 0) {
 			ret = num;
 			dev_err(cpu_dev, "no OPP table is found: %d\n", ret);
@@ -402,13 +406,14 @@ soc_opp_out:
 	if (ret > 0)
 		transition_latency += ret * 1000;
 
+	mutex_init(&set_cpufreq_lock);
+
 	ret = cpufreq_register_driver(&imx6q_cpufreq_driver);
 	if (ret) {
 		dev_err(cpu_dev, "failed register driver: %d\n", ret);
 		goto free_freq_table;
 	}
 
-	mutex_init(&set_cpufreq_lock);
 	register_pm_notifier(&imx6_cpufreq_pm_notifier);
 
 	of_node_put(np);
-- 
1.7.5.4

